File: frontend/src/lib/api/geneve.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface GeneveInterface {
  name: string;
  type: string;
  addresses: string[];
  description: string | null;
  vrf: string | null;
  mtu: string | null;
  mac: string | null;
  disable: boolean | null;
  // GENEVE-specific
  remote: string | null;
  vni: string | null;
  port: string | null;
  // Parameters: IP
  parameters_ip_df: string | null;
  parameters_ip_tos: string | null;
  parameters_ip_ttl: string | null;
  parameters_ip_innerproto: boolean | null;
  // Parameters: IPv6
  parameters_ipv6_flowlabel: string | null;
  // IP settings
  ip_adjust_mss: string | null;
  ip_arp_cache_timeout: string | null;
  ip_disable_arp_filter: boolean | null;
  ip_disable_forwarding: boolean | null;
  ip_enable_arp_accept: boolean | null;
  ip_enable_arp_announce: boolean | null;
  ip_enable_arp_ignore: boolean | null;
  ip_enable_directed_broadcast: boolean | null;
  ip_enable_proxy_arp: boolean | null;
  ip_proxy_arp_pvlan: boolean | null;
  ip_source_validation: string | null;
  // IPv6 settings
  ipv6_accept_dad: string | null;
  ipv6_adjust_mss: string | null;
  ipv6_base_reachable_time: string | null;
  ipv6_disable_forwarding: boolean | null;
  ipv6_dup_addr_detect_transmits: string | null;
  ipv6_source_validation: string | null;
  ipv6_address_autoconf: boolean | null;
  ipv6_address_eui64: string[];
  ipv6_address_no_default_link_local: boolean | null;
  ipv6_address_interface_identifier: string | null;
  // Mirror
  mirror_ingress: string | null;
  mirror_egress: string | null;
  // Redirect
  redirect: string | null;
}

export interface GeneveConfigResponse {
  interfaces: GeneveInterface[];
  total: number;
  by_type: Record<string, number>;
  by_vrf: Record<string, number>;
}

export interface GeneveCapabilities {
  version: string;
  features: Record<string, { supported: boolean; description: string }>;
  version_info?: {
    is_1_4: boolean;
    is_1_5: boolean;
  };
  instance_name?: string;
  instance_id?: string;
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

export interface GeneveBatchOperation {
  op: string;
  value?: string;
}

// ============================================================================
// API Service
// ============================================================================

class GeneveService {
  async getCapabilities(): Promise<GeneveCapabilities> {
    return apiClient.get<GeneveCapabilities>("/vyos/geneve/capabilities");
  }

  async getConfig(refresh: boolean = false): Promise<GeneveConfigResponse> {
    return apiClient.get<GeneveConfigResponse>("/vyos/geneve/config", {
      refresh: refresh.toString(),
    });
  }

  async refreshConfig(): Promise<VyOSResponse> {
    return apiClient.post("/vyos/config/refresh");
  }

  async batchConfigure(
    interfaceName: string,
    operations: GeneveBatchOperation[]
  ): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/geneve/batch", {
      interface: interfaceName,
      operations,
    });
    await this.refreshConfig();
    return result;
  }

  async createInterface(config: {
    name: string;
    description?: string;
    addresses?: string[];
    mtu?: string;
    vrf?: string;
    disabled?: boolean;
    mac?: string;
    remote?: string;
    vni?: string;
    port?: string;
    parameters_ip_df?: string;
    parameters_ip_tos?: string;
    parameters_ip_ttl?: string;
    parameters_ip_innerproto?: boolean;
    parameters_ipv6_flowlabel?: string;
    ip_adjust_mss?: string;
    ip_arp_cache_timeout?: string;
    ip_disable_arp_filter?: boolean;
    ip_disable_forwarding?: boolean;
    ip_enable_arp_accept?: boolean;
    ip_enable_arp_announce?: boolean;
    ip_enable_arp_ignore?: boolean;
    ip_enable_directed_broadcast?: boolean;
    ip_enable_proxy_arp?: boolean;
    ip_proxy_arp_pvlan?: boolean;
    ip_source_validation?: string;
    ipv6_accept_dad?: string;
    ipv6_adjust_mss?: string;
    ipv6_base_reachable_time?: string;
    ipv6_disable_forwarding?: boolean;
    ipv6_dup_addr_detect_transmits?: string;
    ipv6_source_validation?: string;
    ipv6_address_autoconf?: boolean;
    ipv6_address_eui64?: string[];
    ipv6_address_no_default_link_local?: boolean;
    ipv6_address_interface_identifier?: string;
    mirror_ingress?: string;
    mirror_egress?: string;
    redirect?: string;
  }): Promise<VyOSResponse> {
    const operations: GeneveBatchOperation[] = [];

    // GENEVE-specific
    if (config.remote) operations.push({ op: "set_remote", value: config.remote });
    if (config.vni) operations.push({ op: "set_vni", value: config.vni });
    if (config.port) operations.push({ op: "set_port", value: config.port });

    // Basic
    if (config.description) operations.push({ op: "set_interface_description", value: config.description });
    if (config.addresses) {
      for (const addr of config.addresses) {
        operations.push({ op: "set_interface_address", value: addr });
      }
    }
    if (config.mtu) operations.push({ op: "set_interface_mtu", value: config.mtu });
    if (config.vrf) operations.push({ op: "set_interface_vrf", value: config.vrf });
    if (config.disabled) operations.push({ op: "set_interface_disable" });
    if (config.mac) operations.push({ op: "set_mac", value: config.mac });

    // Tunnel parameters
    if (config.parameters_ip_df) operations.push({ op: "set_parameters_ip_df", value: config.parameters_ip_df });
    if (config.parameters_ip_tos) operations.push({ op: "set_parameters_ip_tos", value: config.parameters_ip_tos });
    if (config.parameters_ip_ttl) operations.push({ op: "set_parameters_ip_ttl", value: config.parameters_ip_ttl });
    if (config.parameters_ip_innerproto) operations.push({ op: "set_parameters_ip_innerproto" });
    if (config.parameters_ipv6_flowlabel) operations.push({ op: "set_parameters_ipv6_flowlabel", value: config.parameters_ipv6_flowlabel });

    // IP settings
    if (config.ip_adjust_mss) operations.push({ op: "set_ip_adjust_mss", value: config.ip_adjust_mss });
    if (config.ip_arp_cache_timeout) operations.push({ op: "set_ip_arp_cache_timeout", value: config.ip_arp_cache_timeout });
    if (config.ip_disable_arp_filter) operations.push({ op: "set_ip_disable_arp_filter" });
    if (config.ip_disable_forwarding) operations.push({ op: "set_ip_disable_forwarding" });
    if (config.ip_enable_arp_accept) operations.push({ op: "set_ip_enable_arp_accept" });
    if (config.ip_enable_arp_announce) operations.push({ op: "set_ip_enable_arp_announce" });
    if (config.ip_enable_arp_ignore) operations.push({ op: "set_ip_enable_arp_ignore" });
    if (config.ip_enable_directed_broadcast) operations.push({ op: "set_ip_enable_directed_broadcast" });
    if (config.ip_enable_proxy_arp) operations.push({ op: "set_ip_enable_proxy_arp" });
    if (config.ip_proxy_arp_pvlan) operations.push({ op: "set_ip_proxy_arp_pvlan" });
    if (config.ip_source_validation) operations.push({ op: "set_ip_source_validation", value: config.ip_source_validation });

    // IPv6 settings
    if (config.ipv6_accept_dad) operations.push({ op: "set_ipv6_accept_dad", value: config.ipv6_accept_dad });
    if (config.ipv6_adjust_mss) operations.push({ op: "set_ipv6_adjust_mss", value: config.ipv6_adjust_mss });
    if (config.ipv6_base_reachable_time) operations.push({ op: "set_ipv6_base_reachable_time", value: config.ipv6_base_reachable_time });
    if (config.ipv6_disable_forwarding) operations.push({ op: "set_ipv6_disable_forwarding" });
    if (config.ipv6_dup_addr_detect_transmits) operations.push({ op: "set_ipv6_dup_addr_detect_transmits", value: config.ipv6_dup_addr_detect_transmits });
    if (config.ipv6_source_validation) operations.push({ op: "set_ipv6_source_validation", value: config.ipv6_source_validation });
    if (config.ipv6_address_autoconf) operations.push({ op: "set_ipv6_address_autoconf" });
    if (config.ipv6_address_eui64) {
      for (const prefix of config.ipv6_address_eui64) {
        operations.push({ op: "set_ipv6_address_eui64", value: prefix });
      }
    }
    if (config.ipv6_address_no_default_link_local) operations.push({ op: "set_ipv6_address_no_default_link_local" });
    if (config.ipv6_address_interface_identifier) operations.push({ op: "set_ipv6_address_interface_identifier", value: config.ipv6_address_interface_identifier });

    // Mirror/redirect
    if (config.mirror_ingress) operations.push({ op: "set_mirror_ingress", value: config.mirror_ingress });
    if (config.mirror_egress) operations.push({ op: "set_mirror_egress", value: config.mirror_egress });
    if (config.redirect) operations.push({ op: "set_redirect", value: config.redirect });

    return this.batchConfigure(config.name, operations);
  }

  async updateInterface(
    name: string,
    current: GeneveInterface,
    updated: {
      description?: string | null;
      addresses?: string[];
      mtu?: string | null;
      vrf?: string | null;
      disabled?: boolean | null;
      mac?: string | null;
      remote?: string | null;
      vni?: string | null;
      port?: string | null;
      parameters_ip_df?: string | null;
      parameters_ip_tos?: string | null;
      parameters_ip_ttl?: string | null;
      parameters_ip_innerproto?: boolean | null;
      parameters_ipv6_flowlabel?: string | null;
      ip_adjust_mss?: string | null;
      ip_arp_cache_timeout?: string | null;
      ip_disable_arp_filter?: boolean | null;
      ip_disable_forwarding?: boolean | null;
      ip_enable_arp_accept?: boolean | null;
      ip_enable_arp_announce?: boolean | null;
      ip_enable_arp_ignore?: boolean | null;
      ip_enable_directed_broadcast?: boolean | null;
      ip_enable_proxy_arp?: boolean | null;
      ip_proxy_arp_pvlan?: boolean | null;
      ip_source_validation?: string | null;
      ipv6_accept_dad?: string | null;
      ipv6_adjust_mss?: string | null;
      ipv6_base_reachable_time?: string | null;
      ipv6_disable_forwarding?: boolean | null;
      ipv6_dup_addr_detect_transmits?: string | null;
      ipv6_source_validation?: string | null;
      ipv6_address_autoconf?: boolean | null;
      ipv6_address_eui64?: string[];
      ipv6_address_no_default_link_local?: boolean | null;
      ipv6_address_interface_identifier?: string | null;
      mirror_ingress?: string | null;
      mirror_egress?: string | null;
      redirect?: string | null;
    }
  ): Promise<VyOSResponse> {
    const operations: GeneveBatchOperation[] = [];

    // Simple string fields
    const stringFields: { key: keyof typeof updated; setOp: string; deleteOp: string; currentVal: string | null }[] = [
      { key: "description", setOp: "set_interface_description", deleteOp: "delete_interface_description", currentVal: current.description },
      { key: "mtu", setOp: "set_interface_mtu", deleteOp: "delete_interface_mtu", currentVal: current.mtu },
      { key: "vrf", setOp: "set_interface_vrf", deleteOp: "delete_interface_vrf", currentVal: current.vrf },
      { key: "mac", setOp: "set_mac", deleteOp: "delete_mac", currentVal: current.mac },
      { key: "remote", setOp: "set_remote", deleteOp: "delete_remote", currentVal: current.remote },
      { key: "vni", setOp: "set_vni", deleteOp: "delete_vni", currentVal: current.vni },
      { key: "port", setOp: "set_port", deleteOp: "delete_port", currentVal: current.port },
      { key: "parameters_ip_df", setOp: "set_parameters_ip_df", deleteOp: "delete_parameters_ip_df", currentVal: current.parameters_ip_df },
      { key: "parameters_ip_tos", setOp: "set_parameters_ip_tos", deleteOp: "delete_parameters_ip_tos", currentVal: current.parameters_ip_tos },
      { key: "parameters_ip_ttl", setOp: "set_parameters_ip_ttl", deleteOp: "delete_parameters_ip_ttl", currentVal: current.parameters_ip_ttl },
      { key: "parameters_ipv6_flowlabel", setOp: "set_parameters_ipv6_flowlabel", deleteOp: "delete_parameters_ipv6_flowlabel", currentVal: current.parameters_ipv6_flowlabel },
      { key: "ip_adjust_mss", setOp: "set_ip_adjust_mss", deleteOp: "delete_ip_adjust_mss", currentVal: current.ip_adjust_mss },
      { key: "ip_arp_cache_timeout", setOp: "set_ip_arp_cache_timeout", deleteOp: "delete_ip_arp_cache_timeout", currentVal: current.ip_arp_cache_timeout },
      { key: "ip_source_validation", setOp: "set_ip_source_validation", deleteOp: "delete_ip_source_validation", currentVal: current.ip_source_validation },
      { key: "ipv6_accept_dad", setOp: "set_ipv6_accept_dad", deleteOp: "delete_ipv6_accept_dad", currentVal: current.ipv6_accept_dad },
      { key: "ipv6_adjust_mss", setOp: "set_ipv6_adjust_mss", deleteOp: "delete_ipv6_adjust_mss", currentVal: current.ipv6_adjust_mss },
      { key: "ipv6_base_reachable_time", setOp: "set_ipv6_base_reachable_time", deleteOp: "delete_ipv6_base_reachable_time", currentVal: current.ipv6_base_reachable_time },
      { key: "ipv6_dup_addr_detect_transmits", setOp: "set_ipv6_dup_addr_detect_transmits", deleteOp: "delete_ipv6_dup_addr_detect_transmits", currentVal: current.ipv6_dup_addr_detect_transmits },
      { key: "ipv6_source_validation", setOp: "set_ipv6_source_validation", deleteOp: "delete_ipv6_source_validation", currentVal: current.ipv6_source_validation },
      { key: "ipv6_address_interface_identifier", setOp: "set_ipv6_address_interface_identifier", deleteOp: "delete_ipv6_address_interface_identifier", currentVal: current.ipv6_address_interface_identifier },
      { key: "mirror_ingress", setOp: "set_mirror_ingress", deleteOp: "delete_mirror_ingress", currentVal: current.mirror_ingress },
      { key: "mirror_egress", setOp: "set_mirror_egress", deleteOp: "delete_mirror_egress", currentVal: current.mirror_egress },
      { key: "redirect", setOp: "set_redirect", deleteOp: "delete_redirect", currentVal: current.redirect },
    ];

    for (const field of stringFields) {
      if (field.key in updated) {
        const newVal = updated[field.key] as string | null | undefined;
        if (newVal) {
          operations.push({ op: field.setOp, value: newVal });
        } else if (field.currentVal) {
          operations.push({ op: field.deleteOp });
        }
      }
    }

    // Boolean flags
    const booleanFields: { key: keyof typeof updated; currentKey: keyof GeneveInterface; setOp: string; deleteOp: string }[] = [
      { key: "disabled", currentKey: "disable", setOp: "set_interface_disable", deleteOp: "delete_interface_disable" },
      { key: "parameters_ip_innerproto", currentKey: "parameters_ip_innerproto", setOp: "set_parameters_ip_innerproto", deleteOp: "delete_parameters_ip_innerproto" },
      { key: "ip_disable_arp_filter", currentKey: "ip_disable_arp_filter", setOp: "set_ip_disable_arp_filter", deleteOp: "delete_ip_disable_arp_filter" },
      { key: "ip_disable_forwarding", currentKey: "ip_disable_forwarding", setOp: "set_ip_disable_forwarding", deleteOp: "delete_ip_disable_forwarding" },
      { key: "ip_enable_arp_accept", currentKey: "ip_enable_arp_accept", setOp: "set_ip_enable_arp_accept", deleteOp: "delete_ip_enable_arp_accept" },
      { key: "ip_enable_arp_announce", currentKey: "ip_enable_arp_announce", setOp: "set_ip_enable_arp_announce", deleteOp: "delete_ip_enable_arp_announce" },
      { key: "ip_enable_arp_ignore", currentKey: "ip_enable_arp_ignore", setOp: "set_ip_enable_arp_ignore", deleteOp: "delete_ip_enable_arp_ignore" },
      { key: "ip_enable_directed_broadcast", currentKey: "ip_enable_directed_broadcast", setOp: "set_ip_enable_directed_broadcast", deleteOp: "delete_ip_enable_directed_broadcast" },
      { key: "ip_enable_proxy_arp", currentKey: "ip_enable_proxy_arp", setOp: "set_ip_enable_proxy_arp", deleteOp: "delete_ip_enable_proxy_arp" },
      { key: "ip_proxy_arp_pvlan", currentKey: "ip_proxy_arp_pvlan", setOp: "set_ip_proxy_arp_pvlan", deleteOp: "delete_ip_proxy_arp_pvlan" },
      { key: "ipv6_disable_forwarding", currentKey: "ipv6_disable_forwarding", setOp: "set_ipv6_disable_forwarding", deleteOp: "delete_ipv6_disable_forwarding" },
      { key: "ipv6_address_autoconf", currentKey: "ipv6_address_autoconf", setOp: "set_ipv6_address_autoconf", deleteOp: "delete_ipv6_address_autoconf" },
      { key: "ipv6_address_no_default_link_local", currentKey: "ipv6_address_no_default_link_local", setOp: "set_ipv6_address_no_default_link_local", deleteOp: "delete_ipv6_address_no_default_link_local" },
    ];

    for (const field of booleanFields) {
      if (updated[field.key] !== undefined) {
        const was = (current[field.currentKey] as boolean | null) ?? false;
        const will = (updated[field.key] as boolean | null) ?? false;
        if (will !== was) {
          operations.push({ op: will ? field.setOp : field.deleteOp });
        }
      }
    }

    // Array: addresses
    if (updated.addresses !== undefined) {
      for (const addr of current.addresses) {
        operations.push({ op: "delete_interface_address", value: addr });
      }
      for (const addr of updated.addresses) {
        operations.push({ op: "set_interface_address", value: addr });
      }
    }

    // Array: ipv6 eui64
    if (updated.ipv6_address_eui64 !== undefined) {
      for (const prefix of current.ipv6_address_eui64) {
        operations.push({ op: "delete_ipv6_address_eui64", value: prefix });
      }
      for (const prefix of updated.ipv6_address_eui64) {
        operations.push({ op: "set_ipv6_address_eui64", value: prefix });
      }
    }

    return this.batchConfigure(name, operations);
  }

  async deleteInterface(name: string): Promise<VyOSResponse> {
    return this.batchConfigure(name, [{ op: "delete_interface" }]);
  }
}

export const geneveService = new GeneveService();

File: backend/routers/interfaces/geneve.py
"""
GENEVE Interface Configuration Endpoints

All GENEVE (Generic Network Virtualization Encapsulation) interface endpoints for VyOS configuration.
GENEVE interfaces provide network virtualization tunneling.
"""

import inspect
import logging
from typing import Dict, List, Optional, Any

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field, ConfigDict
from starlette.concurrency import run_in_threadpool

from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
from session_vyos_service import get_session_vyos_service

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/vyos/geneve", tags=["geneve-interface"])


# Stub functions for backwards compatibility with app.py
def set_device_registry(registry):
    """Legacy function - no longer used."""
    pass


def set_configured_device_name(name):
    """Legacy function - no longer used."""
    pass


# ============================================================================
# Request / Response Models
# ============================================================================


class BatchOperation(BaseModel):
    op: str = Field(..., description="Operation name")
    value: Optional[str] = Field(None, description="Operation value (if required)")


class BatchRequest(BaseModel):
    interface: str = Field(..., description="Interface name (e.g., gnv0)")
    operations: List[BatchOperation]


class VyOSResponse(BaseModel):
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


class GeneveInterfaceConfig(BaseModel):
    name: str
    type: str
    addresses: List[str] = Field(default_factory=list)
    description: Optional[str] = None
    vrf: Optional[str] = None
    mtu: Optional[str] = None
    mac: Optional[str] = None
    disable: Optional[bool] = None
    # GENEVE-specific
    remote: Optional[str] = None
    vni: Optional[str] = None
    port: Optional[str] = None
    # Parameters: IP
    parameters_ip_df: Optional[str] = None
    parameters_ip_tos: Optional[str] = None
    parameters_ip_ttl: Optional[str] = None
    parameters_ip_innerproto: Optional[bool] = None
    # Parameters: IPv6
    parameters_ipv6_flowlabel: Optional[str] = None
    # IP settings
    ip_adjust_mss: Optional[str] = None
    ip_arp_cache_timeout: Optional[str] = None
    ip_disable_arp_filter: Optional[bool] = None
    ip_disable_forwarding: Optional[bool] = None
    ip_enable_arp_accept: Optional[bool] = None
    ip_enable_arp_announce: Optional[bool] = None
    ip_enable_arp_ignore: Optional[bool] = None
    ip_enable_directed_broadcast: Optional[bool] = None
    ip_enable_proxy_arp: Optional[bool] = None
    ip_proxy_arp_pvlan: Optional[bool] = None
    ip_source_validation: Optional[str] = None
    # IPv6 settings
    ipv6_accept_dad: Optional[str] = None
    ipv6_adjust_mss: Optional[str] = None
    ipv6_base_reachable_time: Optional[str] = None
    ipv6_disable_forwarding: Optional[bool] = None
    ipv6_dup_addr_detect_transmits: Optional[str] = None
    ipv6_source_validation: Optional[str] = None
    ipv6_address_autoconf: Optional[bool] = None
    ipv6_address_eui64: List[str] = Field(default_factory=list)
    ipv6_address_no_default_link_local: Optional[bool] = None
    ipv6_address_interface_identifier: Optional[str] = None
    # Mirror
    mirror_ingress: Optional[str] = None
    mirror_egress: Optional[str] = None
    # Redirect
    redirect: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)


class GeneveInterfacesConfigResponse(BaseModel):
    interfaces: List[GeneveInterfaceConfig] = Field(default_factory=list)
    total: int = 0
    by_type: Dict[str, int] = Field(default_factory=dict)
    by_vrf: Dict[str, int] = Field(default_factory=dict)


# ============================================================================
# Endpoints
# ============================================================================


@router.get("/capabilities")
async def get_capabilities(request: Request) -> Dict[str, Any]:
    """Return version-aware feature capabilities for geneve interfaces."""
    await require_read_permission(request, FeatureGroup.INTERFACES)
    service = get_session_vyos_service(request)
    from vyos_builders.interfaces.geneve import GeneveInterfaceBuilderMixin
    builder = GeneveInterfaceBuilderMixin(version=service.get_version())
    return builder.get_capabilities()


@router.get("/config", response_model=GeneveInterfacesConfigResponse)
async def get_config(http_request: Request, refresh: bool = False) -> GeneveInterfacesConfigResponse:
    """Get all geneve interface configurations from VyOS."""
    await require_read_permission(http_request, FeatureGroup.INTERFACES)
    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh)
        raw_config = full_config.get("interfaces", {}).get("geneve", {})

        from vyos_mappers.interfaces.geneve_versions import get_geneve_mapper
        mapper = get_geneve_mapper(service.get_version())
        parsed = mapper.parse_interfaces_of_type(raw_config)
        return GeneveInterfacesConfigResponse(**parsed)
    except Exception:
        logger.exception("Unhandled error in get_config")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/batch", response_model=VyOSResponse)
async def batch_configure(http_request: Request, request: BatchRequest) -> VyOSResponse:
    """
    Configure a geneve interface using batch operations.

    **Supported operations (all versions):**
    | Operation | Value | Description |
    |-----------|-------|-------------|
    | `set_interface_description` | Yes | Set description |
    | `delete_interface_description` | No | Remove description |
    | `set_interface_address` | Yes | Add IP address (CIDR) |
    | `delete_interface_address` | Yes | Remove IP address |
    | `set_interface_mtu` | Yes | Set MTU (1200-16000) |
    | `delete_interface_mtu` | No | Reset MTU to default |
    | `set_interface_disable` | No | Administratively disable |
    | `delete_interface_disable` | No | Re-enable interface |
    | `set_interface_vrf` | Yes | Assign to VRF |
    | `delete_interface_vrf` | No | Remove VRF assignment |
    | `set_mac` | Yes | Set MAC address |
    | `delete_mac` | No | Remove MAC address |
    | `set_remote` | Yes | Set tunnel remote address |
    | `delete_remote` | No | Remove tunnel remote address |
    | `set_vni` | Yes | Set Virtual Network Identifier |
    | `delete_vni` | No | Remove VNI |
    | `set_port` | Yes | Set port number |
    | `delete_port` | No | Reset port to default |
    | `set_parameters_ip_df` | Yes | Set DF bit (set/unset/inherit) |
    | `delete_parameters_ip_df` | No | Remove DF setting |
    | `set_parameters_ip_tos` | Yes | Set TOS (0-99) |
    | `delete_parameters_ip_tos` | No | Remove TOS |
    | `set_parameters_ip_ttl` | Yes | Set TTL (0-255) |
    | `delete_parameters_ip_ttl` | No | Remove TTL |
    | `set_parameters_ip_innerproto` | No | Use IPv4 as inner protocol |
    | `delete_parameters_ip_innerproto` | No | Remove inner protocol setting |
    | `set_parameters_ipv6_flowlabel` | Yes | Set IPv6 flow label |
    | `delete_parameters_ipv6_flowlabel` | No | Remove flow label |
    | `set_ip_adjust_mss` | Yes | Adjust TCP MSS |
    | `delete_ip_adjust_mss` | No | Remove MSS adjustment |
    | `set_ip_arp_cache_timeout` | Yes | Set ARP cache timeout |
    | `delete_ip_arp_cache_timeout` | No | Remove ARP cache timeout |
    | `set_ip_disable_arp_filter` | No | Disable ARP filter |
    | `delete_ip_disable_arp_filter` | No | Enable ARP filter |
    | `set_ip_disable_forwarding` | No | Disable IPv4 forwarding |
    | `delete_ip_disable_forwarding` | No | Enable IPv4 forwarding |
    | `set_ip_enable_arp_accept` | No | Enable ARP accept |
    | `delete_ip_enable_arp_accept` | No | Disable ARP accept |
    | `set_ip_enable_arp_announce` | No | Enable ARP announce |
    | `delete_ip_enable_arp_announce` | No | Disable ARP announce |
    | `set_ip_enable_arp_ignore` | No | Enable ARP ignore |
    | `delete_ip_enable_arp_ignore` | No | Disable ARP ignore |
    | `set_ip_enable_directed_broadcast` | No | Enable directed broadcast |
    | `delete_ip_enable_directed_broadcast` | No | Disable directed broadcast |
    | `set_ip_enable_proxy_arp` | No | Enable proxy-arp |
    | `delete_ip_enable_proxy_arp` | No | Disable proxy-arp |
    | `set_ip_proxy_arp_pvlan` | No | Enable private VLAN proxy ARP |
    | `delete_ip_proxy_arp_pvlan` | No | Disable private VLAN proxy ARP |
    | `set_ip_source_validation` | Yes | Source validation (strict/loose/disable) |
    | `delete_ip_source_validation` | No | Remove source validation |
    | `set_ipv6_accept_dad` | Yes | Set DAD mode (0/1/2) |
    | `delete_ipv6_accept_dad` | No | Remove DAD setting |
    | `set_ipv6_adjust_mss` | Yes | Adjust TCP MSS for IPv6 |
    | `delete_ipv6_adjust_mss` | No | Remove IPv6 MSS adjustment |
    | `set_ipv6_base_reachable_time` | Yes | Set base reachable time |
    | `delete_ipv6_base_reachable_time` | No | Remove base reachable time |
    | `set_ipv6_disable_forwarding` | No | Disable IPv6 forwarding |
    | `delete_ipv6_disable_forwarding` | No | Enable IPv6 forwarding |
    | `set_ipv6_dup_addr_detect_transmits` | Yes | Set DAD transmit count |
    | `delete_ipv6_dup_addr_detect_transmits` | No | Remove DAD transmit count |
    | `set_ipv6_source_validation` | Yes | IPv6 source validation |
    | `delete_ipv6_source_validation` | No | Remove IPv6 source validation |
    | `set_ipv6_address_autoconf` | No | Enable SLAAC |
    | `delete_ipv6_address_autoconf` | No | Disable SLAAC |
    | `set_ipv6_address_eui64` | Yes | Add EUI-64 prefix |
    | `delete_ipv6_address_eui64` | Yes | Remove EUI-64 prefix |
    | `set_ipv6_address_no_default_link_local` | No | Remove default link-local |
    | `delete_ipv6_address_no_default_link_local` | No | Restore default link-local |
    | `set_mirror_ingress` | Yes | Mirror ingress to interface |
    | `delete_mirror_ingress` | No | Remove ingress mirror |
    | `set_mirror_egress` | Yes | Mirror egress to interface |
    | `delete_mirror_egress` | No | Remove egress mirror |
    | `set_redirect` | Yes | Redirect incoming packets |
    | `delete_redirect` | No | Remove redirect |
    | `delete_interface` | No | Delete entire interface |

    **VyOS 1.5+ only:**
    | `set_ipv6_address_interface_identifier` | Yes | Set SLAAC interface identifier |
    | `delete_ipv6_address_interface_identifier` | No | Remove SLAAC interface identifier |
    """
    await require_write_permission(http_request, FeatureGroup.INTERFACES)

    try:
        service = get_session_vyos_service(http_request)
        batch = service.create_geneve_batch()

        for op in request.operations:
            if op.op in batch._INTERNAL_BUILDER_METHODS:
                raise HTTPException(
                    status_code=400,
                    detail=f"Operation '{op.op}' is not a valid interface operation",
                )

            method = getattr(batch, op.op, None)
            if method is None:
                raise HTTPException(
                    status_code=400,
                    detail=f"Unsupported operation: {op.op}",
                )

            sig = inspect.signature(method)
            params = [p for p in sig.parameters.keys() if p != "self"]

            if len(params) == 1:
                method(request.interface)
            elif len(params) == 2:
                if op.value is None:
                    raise HTTPException(
                        status_code=400,
                        detail=f"Operation '{op.op}' requires a value",
                    )
                method(request.interface, op.value)
            else:
                raise HTTPException(
                    status_code=400,
                    detail=f"Operation '{op.op}' has unexpected signature",
                )

        response = service.execute_batch(batch)
        return VyOSResponse(
            success=response.status == 200,
            data=response.result if isinstance(response.result, dict) else None,
            error=response.error if response.error else None,
        )
    except HTTPException:
        raise
    except Exception:
        logger.exception("Unhandled error in batch_configure")
        raise HTTPException(status_code=500, detail="Internal server error")
